### PR TITLE
FIX Bug: It can find dependent libraries on its own

### PR DESCRIPTION
**Pull request by KangLin**
See: https://github.com/microsoft/vcpkg/issues/18616

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -280,8 +280,8 @@ if (USE_GNUTLS)
 	endif()
 	target_compile_definitions(datachannel PRIVATE USE_GNUTLS=1)
 	target_compile_definitions(datachannel-static PRIVATE USE_GNUTLS=1)
-	target_link_libraries(datachannel PRIVATE GnuTLS::GnuTLS)
-	target_link_libraries(datachannel-static PRIVATE GnuTLS::GnuTLS)
+	target_link_libraries(datachannel PUBLIC GnuTLS::GnuTLS)
+	target_link_libraries(datachannel-static PUBLIC GnuTLS::GnuTLS)
```
Comment (paullouisageneau) [inline CMakeLists.txt:284]: Why is that change necessary? GnuTLS, OpenSSL and libnice dependencies should not be linked against but not exposed to dependencies for linking.

Comment (paullouisageneau) [inline CMakeLists.txt:284]: Why is that change necessary? GnuTLS, OpenSSL and libnice dependencies should be linked against but not exposed to dependencies for linking, hence the private visibility.

File: cmake/libdatachannel-config.cmake.in [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    find_dependency(GnuTLS)
+else()
+    find_dependency(OpenSSL)
+endif()
+
+#if (@USE_NICE@)
+#    find_dependency(LibNice)
+#else()
+#    find_dependency(LibJuice)
+#endif()
```
Comment (paullouisageneau) [inline cmake/libdatachannel-config.cmake.in:16]: Why is this commented out ? libjuice is always a submodule for now so I guess only libnice should be considered.

Also, what about libSRTP?

Comment (KangLin) [reply cmake/libdatachannel-config.cmake.in:16]: Temporarily comment because it does not export configuration files.

File: CMakeLists.txt (hunk 1 above)
Comment (KangLin) [reply CMakeLists.txt:284]: I thinks GnuTLS, OpenSSL and libnice is same Threads. The application need link they.

Comment (paullouisageneau) [reply CMakeLists.txt:284]: Actually, the application does not need to include GnuTLS headers directories, and only libdatachannel.so needs to link against libgnutls.so, the application does not.

Comment (paullouisageneau) [reply CMakeLists.txt:284]: Threads is a bit different as not only libdatachannel needs to be built with thread flags, the application also needs to. That's why Threads is public.

**Comment by paullouisageneau** [comment]
Could you describe the setup leading to the bug? 

Currently, when I install libdatachannel, the resulting cmake config file includes `INTERFACE_LINK_LIBRARIES "Threads::Threads"`, which looks good to me.

**Comment by paullouisageneau** [comment]
Any more information about this?

**Comment by KangLin** [comment]
I have been busy writing [Rabbit Remote Control](https://github.com/KangLin/RabbitRemoteControl) recently. It will be use your the project 😄 
When I have time, I will write a simple example for this bug.

**Comment by KangLin** [comment]
I wrote a test program [TestLibdatachannel.tar.gz](https://github.com/paullouisageneau/libdatachannel/files/7019679/TestLibdatachannel.tar.gz).
Test Results: 
Dependent library GnuTLS, OpenSSL and Threads do not need to be exported.

**Comment by paullouisageneau** [comment]
Sorry, I still don't get what issue this is supposed to solve.

The main change now seems to be changing Threads to a private dependency. Sure, you don't technically need to have Threads as a public dependency, but then the user needs to call `find_package(Threads)`. What's the point of that change? 

If you use libdatachannel, you need to compile your code with threads support, so you need to link against the special CMake Threads library to add all the necessary compilation and linking flags for that. For instance, on GNU libc, it adds `-pthreads`, which defines `_REENTRANT` so the libc uses thread-local storage for errno and that kind of things.

**Comment by KangLin** [comment]
My application may also use libdatachannel in a single thread. It doesn't need Threads.
If my program uses multi-threading, then of course I need to use find_package(Threads) in the CMakeList.txt of my application to link it.
Of course, This is beyond the scope of this discussion.

If you still think that Threads needs to be public. So come back to this discussion topic.
You need to modify the CMake export configuration file and add find_dependency to Threads dependency. 
In this way, the user does not need to use Find_package(Threads) in his CMakeLists.txt to link it.

**Comment by paullouisageneau** [comment]
> My application may also use libdatachannel in a single thread. It doesn't need link Threads.

This is not possible as libdatachannel will trigger callbacks from different threads.

> You need to modify the CMake export configuration file and add find_dependency(Threads) to link Threads.
> In this way, the user does not need to use Find_package(Threads) in his CMakeLists.txt to link it.

Currently, when you install libdatachannel, the generated cmake export configuration file includes INTERFACE_LINK_LIBRARIES "Threads::Threads". Isn't this correct?

**Comment by paullouisageneau** [comment]
> My application may also use libdatachannel in a single thread.

In practice, this is not possible as libdatachannel will trigger callbacks from a different thread.

> You need to modify the CMake export configuration file and add find_dependency(Threads) to link Threads.
> In this way, the user does not need to use Find_package(Threads) in his CMakeLists.txt to link it.

Indeed, I see the problem. The generated cmake export configuration file includes `INTERFACE_LINK_LIBRARIES "Threads::Threads"`, but CMake does not generate `find_dependency(Threads)`. That kind of defeats the purpose of exporting dependencies.

For the sake of simplicity, it might be indeed better to make Threads private so there is technically no exported dependency.

**Review by paullouisageneau — APPROVED** [review]
